src/lib/Components/ReadMore.tsx

```
@@ -0,0 +1,121 @@
+ * traversing and adds a 'read more' button to the highest text node at that part of the tree.
+ */
+function truncate({
+  root,
+  maxChars,
+  onExpand,
+}: {
+  root: React.ReactNode
+  maxChars: number
+  onExpand(): void
+}): React.ReactNode {
+  // keep track of how many characters we have seen
+  let offset = 0
+  // keep track of how many text nodes deep we are
+  let textDepth = 0
+
+  function traverse(node: React.ReactNode) {
+    if (offset === maxChars) {
+      return null
+    }
+
+    if (typeof node === "string") {
+      let text = node
+      if (node.length > maxChars - offset) {
+        text = node.slice(0, maxChars - offset) as string
+      }
+
+      offset += text.length
+
+      return text
+    }
+
+    if (Array.isArray(node)) {
+      const result = []
+      for (const child of node) {
+        const truncated = traverse(child)
+        if (truncated) {
+          result.push(truncated)
+        }
+        if (offset === maxChars) {
+          return result
+        }
+      }
+      return result
+    }
+
+    if (React.isValidElement(node)) {
+      // TODO: find a way to make the rendering extensible while allowing textDepth to be tracked.
+      // Right now we assume that only these two Text nodes will be used.
+      if (node.type === Sans || node.type === Serif) {
+        textDepth += 1
+      }
+      const children = React.Children.toArray((node.props as any).children)
+      const truncatedChildren = traverse(children)
+
+      if (node.type === Sans || node.type === Serif) {
+        if (textDepth === 1 && maxChars === offset) {
+          truncatedChildren.push([
+            "... ",
+            <LinkText onPress={onExpand}>
+              <Sans size="3" weight="medium">
+                Read more
+              </Sans>
+            </LinkText>,
+          ])
+        }
+        textDepth -= 1
+      }
+
+      return React.cloneElement(node, null, truncatedChildren)
+    }
+  }
```

it just occurred to me that we're not handling nulls, numbers, booleans, and other truthy objects properly. It should probably be:

- null/undefined/true/false - do not get rendered
- arrays - get traversed
- react elements - have their children traversed if present
- everything else - gets coerced to a string

---

(similar question as below). Should we throw an error if it's null/undefined/true/false? Or just return null/undefined?